{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c368a6fa",
   "metadata": {},
   "source": [
    "# Preprocessing WildType Sequences with Point Mutations\n",
    "\n",
    "## Sample\n",
    "\n",
    "Load the varient scores and codon counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "00fabe57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>variant_call_support</th>\n",
       "      <th>codon_substitutions</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>n_codon_substitutions</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>pool1</td>\n",
       "      <td>AAAAAAAAAAAGGAGA</td>\n",
       "      <td>4</td>\n",
       "      <td>GGT166ATG</td>\n",
       "      <td>G166M</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>pool1</td>\n",
       "      <td>AAAAAAAAAAATTTAA</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>pool1</td>\n",
       "      <td>AAAAAAAAAACGCGTA</td>\n",
       "      <td>3</td>\n",
       "      <td>GAA154ACT</td>\n",
       "      <td>E154T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>pool1</td>\n",
       "      <td>AAAAAAAAAACTCCAA</td>\n",
       "      <td>2</td>\n",
       "      <td>TTT156ATG</td>\n",
       "      <td>F156M</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>pool1</td>\n",
       "      <td>AAAAAAAAACCGATTA</td>\n",
       "      <td>2</td>\n",
       "      <td>CAG84GAA</td>\n",
       "      <td>Q84E</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       target library           barcode  variant_call_support  \\\n",
       "0  Wuhan_Hu_1   pool1  AAAAAAAAAAAGGAGA                     4   \n",
       "1  Wuhan_Hu_1   pool1  AAAAAAAAAAATTTAA                     4   \n",
       "2  Wuhan_Hu_1   pool1  AAAAAAAAAACGCGTA                     3   \n",
       "3  Wuhan_Hu_1   pool1  AAAAAAAAAACTCCAA                     2   \n",
       "4  Wuhan_Hu_1   pool1  AAAAAAAAACCGATTA                     2   \n",
       "\n",
       "  codon_substitutions aa_substitutions  n_codon_substitutions  \\\n",
       "0           GGT166ATG            G166M                      1   \n",
       "1                 NaN              NaN                      0   \n",
       "2           GAA154ACT            E154T                      1   \n",
       "3           TTT156ATG            F156M                      1   \n",
       "4            CAG84GAA             Q84E                      1   \n",
       "\n",
       "   n_aa_substitutions  \n",
       "0                   1  \n",
       "1                   0  \n",
       "2                   1  \n",
       "3                   1  \n",
       "4                   1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "current_dir =  os.getcwd()\n",
    "data_dir = os.path.join(current_dir, '..','data', 'processed','results')\n",
    "variant_scores_path = os.path.join(data_dir, 'final_variant_scores', 'final_variant_scores.csv')\n",
    "codon_counts_path = os.path.join(data_dir, 'variants','codon_variant_table_Wuhan_Hu_1.csv')\n",
    "\n",
    "df_variant_scores = pd.read_csv(variant_scores_path)\n",
    "\n",
    "df_codon_counts = pd.read_csv(codon_counts_path)\n",
    "\n",
    "df_codon_counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "431eb526",
   "metadata": {},
   "source": [
    "Our column of interest is ```aa_substitutions```. The way the substituitions are encoded is original amino acid | position in sequence | mutated amino acid. To improve overall context for the transformer model, we will subsitute those mutations into the entire amino acid wildtype sequence. Additionally, multiple mutations in one sequence are encoded as a string with a space between each mutation, i.e: ```'C6F S29H S36D'``` like below. Our next column of interest below is ```'bind'``` which is the binding affinity of that sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b238761",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>wildtype</th>\n",
       "      <th>position</th>\n",
       "      <th>mutant</th>\n",
       "      <th>mutation</th>\n",
       "      <th>bind</th>\n",
       "      <th>delta_bind</th>\n",
       "      <th>n_bc_bind</th>\n",
       "      <th>n_libs_bind</th>\n",
       "      <th>bind_rep1</th>\n",
       "      <th>bind_rep2</th>\n",
       "      <th>bind_rep3</th>\n",
       "      <th>expr</th>\n",
       "      <th>delta_expr</th>\n",
       "      <th>n_bc_expr</th>\n",
       "      <th>n_libs_expr</th>\n",
       "      <th>expr_rep1</th>\n",
       "      <th>expr_rep2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta</td>\n",
       "      <td>N</td>\n",
       "      <td>331</td>\n",
       "      <td>A</td>\n",
       "      <td>N331A</td>\n",
       "      <td>9.41007</td>\n",
       "      <td>0.10297</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>9.31807</td>\n",
       "      <td>9.49961</td>\n",
       "      <td>9.41253</td>\n",
       "      <td>9.72294</td>\n",
       "      <td>-0.27844</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>9.63034</td>\n",
       "      <td>9.81554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Beta</td>\n",
       "      <td>N</td>\n",
       "      <td>331</td>\n",
       "      <td>C</td>\n",
       "      <td>N331C</td>\n",
       "      <td>9.11229</td>\n",
       "      <td>-0.19481</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>9.06925</td>\n",
       "      <td>9.09584</td>\n",
       "      <td>9.17180</td>\n",
       "      <td>9.36532</td>\n",
       "      <td>-0.63606</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>9.18919</td>\n",
       "      <td>9.54145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Beta</td>\n",
       "      <td>N</td>\n",
       "      <td>331</td>\n",
       "      <td>D</td>\n",
       "      <td>N331D</td>\n",
       "      <td>9.31717</td>\n",
       "      <td>0.01007</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "      <td>9.18921</td>\n",
       "      <td>9.38256</td>\n",
       "      <td>9.37976</td>\n",
       "      <td>9.77842</td>\n",
       "      <td>-0.22296</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>9.70504</td>\n",
       "      <td>9.85180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Beta</td>\n",
       "      <td>N</td>\n",
       "      <td>331</td>\n",
       "      <td>E</td>\n",
       "      <td>N331E</td>\n",
       "      <td>9.36899</td>\n",
       "      <td>0.06189</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>9.36041</td>\n",
       "      <td>9.42835</td>\n",
       "      <td>9.31821</td>\n",
       "      <td>10.06567</td>\n",
       "      <td>0.06430</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>9.93181</td>\n",
       "      <td>10.19954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Beta</td>\n",
       "      <td>N</td>\n",
       "      <td>331</td>\n",
       "      <td>F</td>\n",
       "      <td>N331F</td>\n",
       "      <td>9.17224</td>\n",
       "      <td>-0.13487</td>\n",
       "      <td>22</td>\n",
       "      <td>3</td>\n",
       "      <td>9.07594</td>\n",
       "      <td>9.17972</td>\n",
       "      <td>9.26104</td>\n",
       "      <td>9.42578</td>\n",
       "      <td>-0.57560</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>9.30185</td>\n",
       "      <td>9.54970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target wildtype  position mutant mutation     bind  delta_bind  n_bc_bind  \\\n",
       "0   Beta        N       331      A    N331A  9.41007     0.10297          6   \n",
       "1   Beta        N       331      C    N331C  9.11229    -0.19481         27   \n",
       "2   Beta        N       331      D    N331D  9.31717     0.01007         23   \n",
       "3   Beta        N       331      E    N331E  9.36899     0.06189         17   \n",
       "4   Beta        N       331      F    N331F  9.17224    -0.13487         22   \n",
       "\n",
       "   n_libs_bind  bind_rep1  bind_rep2  bind_rep3      expr  delta_expr  \\\n",
       "0            3    9.31807    9.49961    9.41253   9.72294    -0.27844   \n",
       "1            3    9.06925    9.09584    9.17180   9.36532    -0.63606   \n",
       "2            3    9.18921    9.38256    9.37976   9.77842    -0.22296   \n",
       "3            3    9.36041    9.42835    9.31821  10.06567     0.06430   \n",
       "4            3    9.07594    9.17972    9.26104   9.42578    -0.57560   \n",
       "\n",
       "   n_bc_expr  n_libs_expr  expr_rep1  expr_rep2  \n",
       "0          3            2    9.63034    9.81554  \n",
       "1         17            2    9.18919    9.54145  \n",
       "2         15            2    9.70504    9.85180  \n",
       "3         10            2    9.93181   10.19954  \n",
       "4         14            2    9.30185    9.54970  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_variant_scores.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8423b25",
   "metadata": {},
   "source": [
    "The wildtype sequence is stored in ```'wildtype_sequence.fasta'``` in the ```data/raw/``` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b8bcc62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('NITNLCPFGEVFNATRFASVYAWNRKRISNCVADYSVLYNSASFSTFKCYGVSP...KST')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load wildtype codon sequence\n",
    "wt_sequence_path = os.path.join(current_dir,'..', 'data', 'raw','wildtype_sequence.fasta')\n",
    "\n",
    "#skip the first line\n",
    "wt_nuc_seq = next(SeqIO.parse(wt_sequence_path, \"fasta\")).seq\n",
    "\n",
    "#translate codon sequence to amino acid sequence\n",
    "wt_aa_seq = wt_nuc_seq.translate(to_stop=True)\n",
    "\n",
    "#visualize\n",
    "wt_aa_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f72e2818",
   "metadata": {},
   "source": [
    "Lets perform a sequence substitution with the example from above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "074f283d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NITNLFPFGEVFNATRFASVYAWNRKRIHNCVADYDVLYNSASFSTFKCYGVSPTKLNDLCFTNVYADSFVIRGDEVRQIAPGQTGKIADYNYKLPDDFTGCVIAWNSNNLDSKVGGNYNYLYRLFRKSNLKPFERDISTEIYQAGSTPCNGVEGFNCYFPLQSYGFQPTNGVGYQPYRVVVLSFELLHAPATVCGPKKST\n"
     ]
    }
   ],
   "source": [
    "#original wt Amino acid sequence\n",
    "seq = list(str(wt_aa_seq))\n",
    "\n",
    "#sample mutations in one sequence\n",
    "mutations = ['C6F', 'S29H', 'S36D']\n",
    "\n",
    "for mut in mutations:\n",
    "    from_aa = mut[0]\n",
    "    to_aa = mut[-1]\n",
    "    pos = int(mut[1:-1]) -1 # the mutations are indexed starting at 1 not 0\n",
    "    assert seq[pos] == from_aa, f\"expected {from_aa} at position {pos+1}, got {seq[pos]}\"\n",
    "    seq[pos] = to_aa\n",
    "\n",
    "print(''.join(seq))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abb321bb",
   "metadata": {},
   "source": [
    "If you look closely at the output, you can see that ```C``` at position 6 has been swapped with ```F```, etc. These will be the sequences we use for modeling. See ```src/utils``` and ```scripts``` for preprocessing details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "28ff59a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>barcode</th>\n",
       "      <th>target</th>\n",
       "      <th>variant_class</th>\n",
       "      <th>aa_substitutions</th>\n",
       "      <th>n_aa_substitutions</th>\n",
       "      <th>TiteSeq_avgcount</th>\n",
       "      <th>log10Ka</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pool1A</td>\n",
       "      <td>AAAAAAAAAAAAAGAA</td>\n",
       "      <td>N501Y</td>\n",
       "      <td>1 nonsynonymous</td>\n",
       "      <td>T46Q</td>\n",
       "      <td>1</td>\n",
       "      <td>1.955927</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pool1A</td>\n",
       "      <td>AAAAAAAAAAAGGAGA</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>1 nonsynonymous</td>\n",
       "      <td>G166M</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pool1A</td>\n",
       "      <td>AAAAAAAAAAATTTAA</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>wildtype</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>36.004828</td>\n",
       "      <td>8.506648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pool1A</td>\n",
       "      <td>AAAAAAAAAACGCGTA</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>1 nonsynonymous</td>\n",
       "      <td>E154T</td>\n",
       "      <td>1</td>\n",
       "      <td>26.968699</td>\n",
       "      <td>8.586023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pool1A</td>\n",
       "      <td>AAAAAAAAAACTCCAA</td>\n",
       "      <td>Wuhan_Hu_1</td>\n",
       "      <td>1 nonsynonymous</td>\n",
       "      <td>F156M</td>\n",
       "      <td>1</td>\n",
       "      <td>40.906723</td>\n",
       "      <td>7.698023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  library           barcode      target    variant_class aa_substitutions  \\\n",
       "0  pool1A  AAAAAAAAAAAAAGAA       N501Y  1 nonsynonymous             T46Q   \n",
       "1  pool1A  AAAAAAAAAAAGGAGA  Wuhan_Hu_1  1 nonsynonymous            G166M   \n",
       "2  pool1A  AAAAAAAAAAATTTAA  Wuhan_Hu_1         wildtype              NaN   \n",
       "3  pool1A  AAAAAAAAAACGCGTA  Wuhan_Hu_1  1 nonsynonymous            E154T   \n",
       "4  pool1A  AAAAAAAAAACTCCAA  Wuhan_Hu_1  1 nonsynonymous            F156M   \n",
       "\n",
       "   n_aa_substitutions  TiteSeq_avgcount   log10Ka  \n",
       "0                   1          1.955927       NaN  \n",
       "1                   1          0.000000       NaN  \n",
       "2                   0         36.004828  8.506648  \n",
       "3                   1         26.968699  8.586023  \n",
       "4                   1         40.906723  7.698023  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_binding_file = os.path.join(data_dir, 'binding_Kd', 'bc_binding.csv')\n",
    "\n",
    "bc_binding = pd.read_csv(bc_binding_file)\n",
    "\n",
    "bc_binding.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23262034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "aa_substitutions\n",
       "NaN           134261\n",
       "V115L            382\n",
       "V32L             328\n",
       "V115S            316\n",
       "V115R            296\n",
       "               ...  \n",
       "S29Y A81D          1\n",
       "T15A N120R         1\n",
       "N40W Q144W         1\n",
       "A14R L111C         1\n",
       "P54D L183H         1\n",
       "Name: count, Length: 47254, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_binding['aa_substitutions'].value_counts(dropna=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cd4b3b3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
